src/theory/strings/regexp_operation.cpp

```diff
@@ -626,30 +627,38 @@ Node RegExpOpr::derivativeSingle( Node r, CVC4::String c ) {
   return retNode;
 }
 
-void RegExpOpr::firstChars( Node r, std::set<unsigned char> &pcset, SetNodes &pvset ) {
+void RegExpOpr::firstChars(Node r, std::set<unsigned> &pcset, SetNodes &pvset)
+{
   Trace("regexp-fset") << "Start FSET(" << mkString(r) << ")" << std::endl;
-  std::map< Node, std::pair< std::set<unsigned char>, SetNodes > >::const_iterator itr = d_fset_cache.find(r);
+  std::map<Node, std::pair<std::set<unsigned>, SetNodes> >::const_iterator itr =
+      d_fset_cache.find(r);
   if(itr != d_fset_cache.end()) {
     pcset.insert((itr->second).first.begin(), (itr->second).first.end());
     pvset.insert((itr->second).second.begin(), (itr->second).second.end());
   } else {
-    std::set<unsigned char> cset;
+    // cset is code points
+    std::set<unsigned> cset;
     SetNodes vset;
     int k = r.getKind();
     switch( k ) {
       case kind::REGEXP_EMPTY: {
         break;
       }
       case kind::REGEXP_SIGMA: {
-        for(unsigned char i='\0'; i<=d_lastchar; i++) {
+        for (unsigned i = 0; i <= d_lastchar; i++)
+        {
           cset.insert(i);
         }
         break;
       }
       case kind::REGEXP_RANGE: {
-        unsigned char a = r[0].getConst<String>().getFirstChar();
-        unsigned char b = r[1].getConst<String>().getFirstChar();
-        for(unsigned char c=a; c<=b; c++) {
+        unsigned a = r[0].getConst<String>().front();
+        a = String::convertUnsignedIntToCode(a);
+        unsigned b = r[1].getConst<String>().front();
+        b = String::convertUnsignedIntToCode(b);
+        Assert(a < b);
```

Comment:
I think this would be a bit excessive as an assertion. We could test other cases where we'd expect things to be rewritten, but the code does the right thing. I guess the difference in this assertion appears to be that we expect the range to be non-empty (which is also checked in the type rules).

Reply:
Why would this be excessive?
IMO an assertion is only then redundant, when you have something like this:
```
if (a < b)
{
  ...
  Assert (a < b);
  ...
}
```
In any other case, it certainly does not hurt and further serves as documentation.

Reply:
The reason I think its excessive is that I could list all the cases that I think my rewriter should rewrite, e.g:
```
Assert( !P1(a,b) ); // since my rewriter ensures that re.range(a,b) is rewritten when P1(a,b) holds))
...
Assert( !Pn(a,b) ); //since my rewriter ensures that re.range(a,b) is rewritten when Pn(a,b) holds))
```
It would help if the code below depended upon a particular property, but the property Andres mentioned is not critical.

Reply:
OK

Reply:
Wait, actually this loop can still be an infinite loop (that's actually why I was suggesting the assertion): If you have a range from 0 to max int, you have an infinite loop.

Reply:
Ah ok, that makes sense. I'll update.  (Let's hope we never deal with alphabets that large with this code haha).